### ISSUE TITLE
[Bug] Length Tool (and possibly other tools) fire ANNOTATION_MODIFIED events when it shouldn't

### ISSUE THREAD

**Issue by mcullifer** [cornerstonejs/cornerstone3D#939]
### Describe the Bug

I'm trying to set up a system where if an annotation is **created** then it calls an api endpoint for adding that annotation to a database. If an existing annotation is **modified** it calls an api endpoint for updating an existing record for it in the database.

The bug I've noticed is with the design of the Length Tool (and this probably is happening with the other annotation tools), an empty annotation is added to the `AnnotationManager` as soon as you begin drawing the annotation. This means that `ANNOTATION_MODIFIED` event is fired constantly all the way until the annotation is completed, then `ANNOTATION_COMPLETED` is fired. This means if you are creating a brand new Length annotation the order of events is
1. `ANNOTATION_ADDED`
... `ANNOTATION_MODIFIED` (for however long you're dragging it)
N. `ANNOTATION_COMPLETED`

This is a design flaw because now there's no realistic way for me to gauge when a brand new annotation is created so I can call an endpoint for **adding** a new annotation to the database vs **updating** and existing record. My thought process is that if an existing annotation in `AnnotationManager` is modified then `ANNOTATION_MODIFIED` should fire. If a brand new annotation is being created then **ONLY** `ANNOTATION_CREATED`, `ANNOTATION_ADDED` should be called. Since all 3 are called in the process of creating an annotation there's no way to tell and I'm forced to hit my api 2 times and 1 of those calls will certainly fail (when it's hitting the update endpoint for a record that doesn't exist).

The issue boils down to the tool adding an empty annotation as soon as you start dragging with the `points` set to [0,0,0]. This shouldn't happen. An annotation can't be "modified" if you haven't even finished drawing the annotation yet. The only way around this is extending the `LengthTool` class and overriding `_endCallback()` to be a new function like
```ts
class LengthTool extends CornerstoneLengthTool {
  originalEndCallback: (evt: InteractionEventType) => void;

  constructor (....) {
    super(...);
    this.originalEndCallback = this._endCallback;
    this._endCallback = onDragEnd;
  }

  onDragEnd(evt: InteractionEventType) {
     let annotation = this.editData?.annotation;
     // Do whatever you need to do here
     this.originalEndCallback(evt);
  }
}
```
which is cumbersome to do for every single tool that this happens in.

### Steps to Reproduce

1. Add event listeners for `ANNOTATION_ADDED`, `ANNOTATION_MODIFIED`, and `ANNOTATION_COMPLETED`
2. Slowly start drawing a length annotation
3. View the console to see that all 3 of the events are fired

### The current behavior

All 3 events are fired when creating a new annotation

### The expected behavior

Only `ANNOTATION_ADDED` and `ANNOTATION_COMPLETED` should be called when first creating an annotation.

### OS

Windows 11

### Node version

18

### Browser

1.61.101 Chromium: 120.0.6099.71 (Official Build) (64-bit)

**Comment by abustany** [comment]
I can imagine various scenarios where you want to be notified of an annotation being modified, no matter if it's a "new" or an "existing" one... We have a use case similar to what you describe at https://rapmed.net, and there I just listen to `ANNOTATION_COMPLETED`. Is there a reason you can't just ignore the `ANNOTATION_MODIFIED` event?

**Comment by mcullifer** [comment]
> I can imagine various scenarios where you want to be notified of an annotation being modified, no matter if it's a "new" or an "existing" one... We have a use case similar to what you describe at https://rapmed.net, and there I just listen to `ANNOTATION_COMPLETED`. Is there a reason you can't just ignore the `ANNOTATION_MODIFIED` event?

`ANNOTATION_COMPLETED`, at least on the length tool, is only called if it is a new annotation [see here](https://github.com/cornerstonejs/cornerstone3D/blob/81e281d066bc2d92c2b7eb7cebbaae49da439fa8/packages/tools/src/tools/annotation/LengthTool.ts#L371-L379). So the only way for me to get access to the state of the annotation when it is the last "modified" is to do the method override like I showed in the post. After `_endCallback()` is fired it sets `this.editData = null`. If I subscribe to `ANNOTATION_MODIFIED` it will fire as long as any modifications are happening and there's no way for me to tell if it's the **final** and true result after a modification is complete. My post is about `ANNOTATION_MODIFIED` gets called before it is complete it's while it's being modified. I need access to the annotation data in 2 cases
1. A new annotation is created
2. An existing annotation is modified (and has been finished being modified)
Doing a `setTimeout()` on the `ANNOTATION_MODIFIED` doesn't work because `ANNOTATION_MODIFIED` fires when a brand new annotation is being created and before the end dot is placed.

**Comment by abustany** [comment]
OK, I understand your problem. In our case, "saving" is an explicit user action so we somehow don't have this problem. Maybe a new `ANNOTATION_MODIFICATION_COMPLETED` event would help there :shrug:

**Comment by mcullifer** [comment]
Yeah if that event type did exist it would probably fit in an `else {}` right after [here](https://github.com/cornerstonejs/cornerstone3D/blob/81e281d066bc2d92c2b7eb7cebbaae49da439fa8/packages/tools/src/tools/annotation/LengthTool.ts#L371-L379) when it checks `if (newAnnotation) {}`

**Comment by sedghi** [comment]
Hmm, it seems like we can't remove the `annotation_added` event with current architecture, but perhaps we can create an additional event that triggers inside the endCallback that is like ANNOTATION_MODIFIED_COMPLETED. What are your thoughts?

**Comment by sedghi** [comment]
@mcullifer If you like, you can create that PR and I will review it. Seems to be small chnage

**Comment by mcullifer** [comment]
@sedghi After experimenting some more it seems like the issue I'm facing is more of a race condition issue. I don't think adding the `ANNOTATION_MODIFIED_COMPLETED` event will fully solve it. I noticed that the `_calculateCachedStats()` function is ran on the next animation frame. When you console log out the cached stats after drawing the line quickly the `length` property is not the same as the label that's in the viewport. When you draw the line slowly it typically will match. I was able to get around this by wrapping the `_calculateCachedStats()` function in a public method and calling that on the subscribers side.

I can still make a PR for adding the else clause to fire a new `ANNOTATION_MODIFIED_COMPLETED` event, but I'm not sure if it's all that useful do to this edge case. I'm also not familiar enough with cornerstone's codebase to know if there's a reason why the cached stats need to be calculated on the next animation frame instead of inside of `_endCallback`. It has a dependency on the `IEnabledElement`, but I think all of this should be accessible at any time? Not sure, what are your thoughts?

**Comment by sedghi** [comment]
race condition on what? Are you saying there is an extra calculate cached stats getting called after the mouse up? I don't understand what is the race between

**Comment by mcullifer** [comment]
Hm maybe it's related to that throttling. Basically with the code I have in the original post I get the annotation data object out, then I call the original _endCallback. Afterwards I use the EventTarget to dispatch a custom event with the annotation. When I did it this way I noticed that depending on how fast I drew the annotation the length property wouldn't be correct. If I drew it slowly it would be correct.

I then tried to just dispatch the annotationUID instead of the whole annotation and then used the FrameOfReferenceSpecificAnnotationManager to get the annotation based on the annotationUID. This also had an incorrect length property in cachedStats. When I did an intentional 100ms timeout it usually had the correct cachedStats with a length that matches what's in the viewport. This behavior led me to believe there was a race happening between me trying to access the annotations cachedStat length value before it did that calculation again (or before the animation frame ran). I decided to manually expose the calculateCachedStats function and call it again myself to ensure I'm getting the correct value every time and this seemed to have a 100% consistent result.

This behavior makes me think that dispatching an ANNOTATION_MODIFIED_END event would have the same issue. Even if I dispatched my custom event after the original _endCallback ran it still was out of date if I drew fast enough.

**Comment by sedghi** [comment]
I understand! If you disable the throttle, I'm confident that we can eliminate any inconsistencies. Perhaps we should consider firing the `annotation_modified_end` within the `renderAnnotation` function. This way, we can ensure that it occurs during the animation frame, and even with throttling, we can guarantee the most up-to-date information.
